src/hss/reference_impl_private_key.rs

```diff
@@ -1,34 +1,102 @@
 use crate::{
     constants::{
-        LmsTreeIdentifier, Seed, D_TOPSEED, ILEN, LMS_LEAF_IDENTIFIERS_SIZE,
-        MAX_ALLOWED_HSS_LEVELS, MAX_HASH_SIZE, REFERENCE_IMPL_PRIVATE_KEY_SIZE, SEED_CHILD_SEED,
-        SEED_LEN, SEED_SIGNATURE_RANDOMIZER_SEED, TOPSEED_D, TOPSEED_LEN, TOPSEED_SEED,
-        TOPSEED_WHICH,
+        LmsTreeIdentifier, D_TOPSEED, ILEN, LMS_LEAF_IDENTIFIERS_SIZE, MAX_ALLOWED_HSS_LEVELS,
+        MAX_HASH_SIZE, MAX_SEED_LEN, REF_IMPL_MAX_PRIVATE_KEY_SIZE, SEED_CHILD_SEED,
+        SEED_SIGNATURE_RANDOMIZER_SEED, TOPSEED_D, TOPSEED_LEN, TOPSEED_SEED, TOPSEED_WHICH,
     },
     hasher::HashChain,
     hss::{definitions::HssPrivateKey, seed_derive::SeedDerive},
     util::helper::read_and_advance,
     HssParameter, LmotsAlgorithm, LmsAlgorithm,
 };
 
-use core::{convert::TryInto, marker::PhantomData, mem::size_of};
+use core::{convert::TryFrom, convert::TryInto, marker::PhantomData};
 use tinyvec::ArrayVec;
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct Seed<H: HashChain> {
+    data: ArrayVec<[u8; MAX_SEED_LEN]>,
+    phantom: PhantomData<H>,
+}
+
+impl<H: HashChain> Seed<H> {
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        H::OUTPUT_SIZE as usize
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [u8] {
+        self.data.as_mut_slice()
+    }
```

Comment:
Just as a note:
Using the `Deref` & `DerefMut` trait should work here as well and would have saved some changes in the code (see in [Hasher for an example](https://github.com/Fraunhofer-AISEC/hbs-lms-rust/blob/master/src/hasher/mod.rs#L18-L23)

But as this works as well, let's keep it like this.

Reply:
Ah, I just have seen you have used it below, but commented it out.

Just out of interest: Would it from a technical point not work wit `Deref`?

Reply:
It almost worked, however, due to automatic dereference in Rust, it caused issues with method signatures expecting &Seed, but receiving &ArrayVec<...> instead

Reply:
Ah, makes sense :ok_hand: